### DISCUSSION TITLE
Scientific simulation in Bevy

### DISCUSSION THREAD

**Discussion by alice-i-cecile** [bevyengine/bevy#1678] [Ideas]
Bevy has tremendous potential as a framework for scientific simulations: it's fast, modular, easy to read/write and hooks in readily to convenient features like graphics and UIs.

This is a tracking issue to discuss the important features to help us get there (h/t @alec-deason). Many of these are "ecosystem" issues, rather than something that needs to be solved by the Bevy engine itself.

## Universal features

1. Improved documentation and API maturity. Templates and guides are a big part of this, as is battle-tested correctness.
2. Charts / graphing integration. See https://github.com/eliotbo/bevy_plot for a community-made option!
3. Statistical integrations.
4. Ergonomic data logging.
5. Headless mode (#22, #436), with good file-driven configuration.
6. Deterministic RNG. This includes good enough APIs to get to zero-system ambiguities (#1312) and ordered query iteration (#1470), but the story is [complex](https://discord.com/channels/691052431525675048/692648082499829760/821814745824624670).
7. Fixed point support, for more stable computations (see #1680). 
8. Ergonomic saving and loading (#1442) with checkpointing.
9. Cross-machine distributed processing.
10. System order visualization and modification.
11. Better testing and correctness tools. #1481 is a good start.

## Domain specific features
1. Indexes (#1587, #1527), for simple and correct grid-based simulations.
2. (Partial) differential equations (PDE) integrations. 
3. Linear algebra integrations.
4. Machine learning integrations.
5. Native physics functionality.

**Comment by alec-deason** [comment]
It seems like there are at least two distinct use cases that could be interesting to think about. One is where bevy directly runs a scientific simulation, which seems to be what the description above is focused on. But there's also clearly a need for test bed simulations to use in robotics and other similar domains. This has been mentioned at least once in #1207. I assume that that use case has softer requirements around mathematical provability and reproducibility and is more like a game with the player out of the loop.

Related to that are the AI test bed setups that various people have built, either on top of existing games like StarCraft or using custom engines. It seems like the main missing feature to enable that is a way for the external AI tool to introspect game state, either by looking at a rendered view as in #1207 or somehow introspecting the ECS store directly. That ties in to other conversations about modding where the ability of non-bevy, or even non-rust, code to interact with the ECS is interesting.

**Reply by alice-i-cecile** [reply]
A third related use case: CAD software (ping @aevyrie). In effect, you're combining a scientific simulation with a classic GUI in a seamless fashion, and want very configurable controls.

**Reply by alice-i-cecile** [reply]
I would posit that "scientific simulation as a tool for producing a paper" and "scientific simulation as a tool for understanding and decision making" are actually two distinct but related user stories. In the former, you want to be able to run it non-interactively and get out some pretty pictures and graphs. In the latter, you end up with something much closer to a classical game, where you want polish and user interaction and experimentation with various simulation parameters and interventions.

**Reply by alec-deason** [reply]
Hmm. My experience with this, which admittedly is limited to one very specific academic environment, is that there isn't really a distinction. The researchers just fiddle with the tool as interactively as it will allow until they have something that works the way they want and then the paper is based on that frozen version of the interactive thing. There isn't really a separate "for publication" version because you have to do the exploratory work even for that.

People talk about polished user interaction for exploratory work but in practice the best version of that seems to be RStudio or Jupyter, as much as I shudder to my core to admit that. But just having a very general purpose calculator that is at least _slightly_ non-cryptic is a win.

**Reply by alice-i-cecile** [reply]
The two places I've seen that latter pattern ("scientific understanding") really clearly is in a) forest management planning b) educational software. Rather than a hacked together interface like notebooks, you end up wanting a real polished app (think RShiny) with knobs you can tweak and interventions you can pick.

For initial twiddling purposes, I think that live parameter tweaking / hot reloading tools ala `bevy_inspector_egui` are going to be the way to go, and can hopefully displace the notebook pattern in a sensible fashion.

**Reply by alec-deason** [reply]
Yeah, that makes sense as a second phase of exploratory stuff. You have the basic model, now you're going to do a lot of experimenting with parameters so it makes sense to have a less flexible but more fiddleable GUI view into it.

It strikes me that the process of exploring model parameters in a scientific context might be very similar to the process of tuning a game late in development. You have the structure of the model down but you need to work out what it's sensitive to, or what interventions have a desired effect. Whether that's species succession outcomes after different tree harvesting patterns or DPS patterns in response to different upgrade drops the parameter tuning task might benefit from the same kind of tooling.

What I'm imagining is something like `bevy_inspector_egui` or as you say some kind of RShiny style GUI toolkit backed by a persistent, version controllable settings system. So the parameters can be loaded from file and also manipulated and saved from the development GUI. I guess that's functionality that an editor would normally provide? But having something that leans into graphing and data visualization seems a bit different and interesting.

**Reply by aevyrie** [reply]
> A third related use case: CAD software (ping @aevyrie). In effect, you're combining a scientific simulation with a classic GUI in a seamless fashion, and want very configurable controls.

CAD could definitely be implemented in either of alec-deason's cases, although that's true for any application. I suspect the majority of projects that don't need the realtime rendering/computing advantages garnered by an ECS will put all their business logic into a separate crate, and only use the bevy-specific features for building an application and UI. This library would presumably handle all the scientific computing in an async fashion.

Forgive the apparent tangent, but where I foresee Bevy really shining is if we develop tools that streamline the process of building an application. A WYSIWYG editor for GUI will definitely be a boon for this, but there are a lot of other things that add drag to the process. When I was building tolstack, these were the "simple" things like file ser/de, a native file explorer, autosaving, application settings, theming, graphing, and exporting simulation results to csv. Even though these aren't the things I would normally associate with scientific simulation, I think they solve a real need for people who want to make polished simulation tools. Maybe a useful comparison is that Bevy _could_ be to scientific simulation tools what dear-imgui is to graphics programming.

**Reply by alec-deason** [reply]
When I think of bevy as a scientific tool I'm thinking of it mostly as a framework for simulation, and that's all ECS. I think you're right that having bevy be a first class application framework would make it interesting to anyone building an application, including a scientific one. But I think ECS oriented, agent based simulation is something that Bevy basically does now that is distinctive and interesting for some kinds of scientific software and if the goal is to make bevy attractive to scientific users (which I don't know that it is, but it would be neat) then you'd lean into that.

**Reply by aevyrie** [reply]
Yeah, you're completely right, that seems like the most natural use of Bevy - the ECS would make agent-based simulation really nice. I think we have totally different use cases in mind within the same(ish) domain 😄. Regardless, I think there are some common needs across the scientific landscape we have a common need for, namely graphing and stats. I saw egui recently added graphing - using that with the inspector tool would be a really seamless way of gleaning statistical insight from simulations interactively.

I find Bevy uniquely attractive from any other application framework for the CAD stuff I want to do because: 3D is first class, it's all in one fast language, I can target native/web, and it's open source. Godot was probably my second option, but I really wanted to do everything in Rust. I hope that gives more context to my earlier comment!

**Comment by alice-i-cecile** [comment]
Another critical domain-specific integration: GIS.

**Reply by j209** [reply]
And also BIM(Building information modeling) construction simulation,Digital twins and so on

**Comment by aristaeus** [comment]
I want charts/graphing! But I'm not sure where to start. Is there an obvious Rust library to integrate, or will we need to roll our own?

**Reply by aevyrie** [reply]
Maybe creating a Bevy backend for `plotters`? https://plotters-rs.github.io/home/#!/

**Reply by alice-i-cecile** [reply]
Long term, I think the tech will be to reuse Bevy's UI framework to create a native graphing library: particularly taking the ideas from #1522 to replicate something with the flavor of the grammar of graphics using components.

That's a long way out though, so I'd definitely start by providing a simple interface to an existing crate.

**Reply by alec-deason** [reply]
Hmm. This seems wrong to me. Graphing is very hard to get right. All the adequate graphing tools (d3, matplotlib, whatever the thing in R is called, etc) are huge, dizzyingly sophisticated tools. Graphing is clearly of use to Bevy users, scientific or not, but it will (I assert) never be a core feature of Bevy so if we tried to do something native it would inevitably be inadequate and, worse, different than whatever becomes the standard for Rust graphing. I think we are far better, even in the very long term, focusing on how to get graphable data out of the ECS and then integrate with whatever the best of breed graphing tool in Rust is.

**Reply by alec-deason** [reply]
That said, someone should implement https://vega.github.io/vega/ in bevy because it's super neat.

**Reply by alice-i-cecile** [reply]
I certainly don't think it should be part of Bevy proper: my vision was for a `bevy_charts` ecosystem crate. The ability to seamlessly integrate your charts into the rest of your GUI is a killer feature for some applications (think Factorio, or forest modelling simulations). Doing so by piping data out of the ECS seems very challenging.

Bevy already needs rendering, interactivity, text styling, UI-style object positioning, scaling, live-updating, complex data pipelines and so on. These crates are huge dizzying monoliths, but almost all of the hard bits are problems that Bevy needs to solve anyways :)

**Reply by alec-deason** [reply]
You may well be right. The GUI integration point is very compelling. My intuition is still that we end up being more useful just for graphing (ignoring deep GUI integration) if something close to bevy provides ergonomic ways to get aggregates and stats out of the ECS and also a canvas in the UI and then an external graphing library consumes the data and draws the charts. That's simply based on how hard it is to build a good graphing library that's flexible enough for general use.

I think I disagree that piping data out of the ECS is a bigger challenge with a loosely integrated tool than a tightly integrated one, though maybe I'm not thinking clearly about the problem. I assume that any graphing problem is going to be structured something like this: collect the results of running a mapping function over one or more queries, optionally do an aggregation on that data, draw a chart of the data. It seems to me that however the graphing tool works, the data it's going to be consuming will be a `Vec<f32>` or something similarly generic which puts a tool that knows nothing about bevy on the same footing as a bevy specific tool, at least from the perspective of data ingestion.

**Reply by alice-i-cecile** [reply]
> I think I disagree that piping data out of the ECS is a bigger challenge with a loosely integrated tool than a tightly integrated one, though maybe I'm not thinking clearly about the problem.

Ah, my wording was unclear here. I was arguing that GUI integration is harder if we're loosely coupled, not that data piping is harder if we're loosely coupled.

I like your ideas here, so I think there are two paths that should follow. The short-term one has two parts:
1. a) Improve Bevy to make it easy to export data from the ECS in RON, CSV and so on. This can be fed into an arbitrary graphing library, used for machine learning, used for logging etc.
b) Create a nice canvas API for quickly and easily drawing arbitrary textures directly to the screen. This probably shares work with folks who want to manually manage pixel art for example.

And then in the long term, once the UI is settled.
2. Use the existing UI tools and paradigms to slowly build up an ecosystem `bevy_charts` crate for deeply integrated use cases and native plotting.

**Reply by alec-deason** [reply]
> Ah, my wording was unclear here. I was arguing that GUI integration is harder if we're loosely coupled, not that data piping is harder if we're loosely coupled.

Oh, I see now. Sorry, I was just being dense.


>     1. a) Improve Bevy to make it easy to export data from the ECS in RON, CSV and so on. This can be fed into an arbitrary graphing library, used for machine learning, used for logging etc.

This is actually not quite what I meant, though it could be very useful. I was advocating for integrating with an in-process Rust graphing library, not shelling out to a totally external tool.

>        b) Create a nice canvas API for quickly and easily drawing arbitrary textures directly to the screen. This probably shares work with folks who want to manually manage pixel art for example.

Yeah, this seems generally useful

**Reply by aevyrie** [reply]
>a) Improve Bevy to make it easy to export data from the ECS in RON, CSV and so on. This can be fed into an arbitrary graphing library, used for machine learning, used for logging etc.

I think a nice intermediate goal is developing plugins that explore ways of aggregating data from the ECS. Because the ECS isn't far removed from a database query language, there might be an ergonomic way to query the ECS directly to pipe out aggregate data. People could build on this to pipe into existing rust graphing crates.

**Reply by alice-i-cecile** [reply]
> This is actually not quite what I meant, though it could be very useful. I was advocating for integrating with an in-process Rust graphing library, not shelling out to a totally external tool.

Ah I see. So I guess you port it out to an in-process vanilla format (e.g. dataframes from [Polars](https://github.com/ritchie46/polars)), and then use *that* to write it out into a stable file format. Okay yeah, that seems much more sane.

**Reply by aevyrie** [reply]
Something like this would be nifty: https://www.flecs.dev/query_demo

**Reply by aristaeus** [reply]
As well as aggregating ECS data, it will be common to build a time series as the game/simulation runs. I currently just have a `Vec<f32>` component on entities and a system to populate it, but it is probably worth thinking about how to make this more ergonomic.

* Time series on entities
* Aggregate time series over several entities.

**Reply by alec-deason** [reply]
Aggregating and down sampling over a single entity's time series is probably also useful.

**Reply by Bauxitedev** [reply]
For basic plotting you can use [egui-bevy](https://github.com/mvlabat/bevy_egui). Here's an example from my program:

![image](https://user-images.githubusercontent.com/31383123/121911254-e7bcd000-cd2f-11eb-8718-48b4ec48b0d4.png)

**Comment by alice-i-cecile** [comment]
Another general feature: easy parallel process replication to enable monte-carlo simulation. See this [thread](https://discord.com/channels/691052431525675048/742884593551802431/822205119143084032) on Discord.

**Reply by CGMossa** [reply]
I've picked up my efforts of using bevy for my projects. At this point I'd love some pointers as to how to do this, in any way possible. Should I make an issue or something similar?

**Reply by alice-i-cecile** [reply]
Make an issue that asks for an example IMO. Please link to both this thread and https://github.com/bevyengine/rfcs/pull/16 as part of that, which may be very useful for enabling this use case.

**Comment by aevyrie** [comment]
@alice-i-cecile and I were chatting about UOM, I am of the opinion that this is a fundamental primitive that Bevy might want to be opinionated on, so that dependent crates can interoperate. The `uom` crate might be a good choice. This will impact scaling (size), time, and physics.

With Bevy and the ECS being so friendly to plugins, being opinionated on units of measure may be critical to avoid ecosystem fracturing with CAD/simulation/game physics plugins.

**Reply by alice-i-cecile** [reply]
Agreed with all of the above. It [appears](https://www.reddit.com/r/rust/comments/ghmkya/choosing_crate_for_typesafe_measurements/) that [`uom`](https://github.com/iliekturtles/uom) is more popular, more mature and better designed than [`measurements`](https://github.com/rust-embedded-community/rust-measurements). My initial inspection of their API lines up with this.

[`dimensioned`](https://github.com/paholg/dimensioned) looks to be the other candidate we should consider, and seems quite similar to `uom`. I'm not sure on their relative strengths, and both seem reasonably active.

**Reply by alice-i-cecile** [reply]
Dimensioned seems somewhat less mature and much less popular: it seems like the major selling point relative to `uom` is no-std, which doesn't matter in our use case. Fairly strongly leaning towards `uom` as a result: interop outside the Bevy ecosystem is very nice too.

Note: this choice also has implications on our choice of rationals; it looks like they're using `num-rational` as the dependency which seems perfectly fine.

**Reply by aristaeus** [reply]
One thing to watch out for is handling vectors with units. `uom` isn't generic over the scalar, so instead of `Length<Vec>` you need to do `Vec<Length>`. That's probably a better API choice, but
a) GLAM types aren't generic over their scalar
b) When I tried doing this with `nalgebra`, it was painful enough that I gave up. I forget the details, but one issue is that SIMD will only work when the vectors are over native types.

**Reply by aevyrie** [reply]
These are good points. A potential solution would be that all internal ops could be optimized glam types, with only externally exposed functions thinly wrapped with `uom` or similar types. We'd have to think carefully about the implications of this, and where it is _truly_ needed. E.g.: the rendering system probably wouldn't benefit from having units on `Transform`s or lighting parameters. However, the physics system might benefit from applying real measurement units to the underlying rendering scalars to prevent physics bugs and make the solver unit agnostic.

**Comment by adsick** [comment]
actually would be very nice to be able  to draw some pointclouds/octrees e.t.c. in Bevy, I already have an application and I need to implement the visualization of the physical process.

**Comment by maciejwnuk** [comment]
I think it would be nice to start off as a plugin (https://github.com/bevyengine/bevy/blob/main/docs/plugins_guidelines.md) and maybe then move on to more "standalone" format.

**Reply by maciejwnuk** [reply]
I would even sign up for collab. 😃

**Comment by RedlineTriad** [comment]
Fixed time-step simulation without a frame-rate limit would also be nice.
Would also be useful for testing to just run bevy with a fixed time-step say 10 ms for 1000 frames.

My use-case would something like https://www.codingame.com/training/easy/mars-lander-episode-1
Where there a various situations and you have to write code to land your rocket.
You could try to simulate it at faster rate than normal, say 1000x but it would be much more elegant to just specify a time-step and then just go as fast as possible.

**Reply by ManevilleF** [reply]
You can specify time step constraints to systems

**Reply by RedlineTriad** [reply]
> You can specify time step constraints to systems

But I don't know of any way to do it with unlimited frame rate, simulate 1000 seconds of physics in 1 second with 10ms time-step.

**Reply by alice-i-cecile** [reply]
A proper frame-rate control API (#1343) should go a long way to solving this need.

**Comment by jueewo** [comment]
I've been working in agent-based modelling and decision support systems and I'm considering to use bevy for future implementations, academically and for an upcoming client project. So right now I'm working on a first prototypic implementation and I'm interested to share experiences.

**Comment by ScottBoyce** [comment]
I know this is a pretty old thread, but I thought I ask if there has been any progress for this with simulations? I've been looking into Rust as an alternative to Fortran for numerical simulation or to provide graphical interface/visualization to the Fortran programs.

**Reply by alice-i-cecile** [reply]
I'm going to be giving a talk at the Scientific Computing in Rust workshop on June 4th, 2025 :) Not about this specifically (it's about lessons from game dev), but it's still an active interest of mine, even now that I'm a maintainer!

Broadly, I would say that Bevy is a solid choice for scientific simulations these days, but the ecosystem support for plotting and statistics is still very weak. I would also like improved determinism tools within Bevy: it's possible but not as smooth as I would  like.

Bevy's most successful commercial users (e.g. Foresight Spatial Labs) right now use it for simulation and CAD, using more or less Bevy to do so :) Definitely possible, even at extreme scales with enough R&D.

Overall opinion right now is: Bevy in 2025 is a great choice for scientific visualization and simulation at small to large scales, but probably only suitable for post-hoc visualization if you have something truly massive like weather simulations or astronomical data crunching. If you need statistics or plotting, you're likely going to need to call out to another language or roll your own solution. There's still not as much documentation or as many out-of-the-box features for this as I would like, but as far as I'm aware, it's not like there's strong competition elsewhere that handles the shared boilerplate either 🤷🏽‍♀️

**Reply by ScottBoyce** [reply]
@alice-i-cecile awesome, I look forward to seeing. Ive been watching the past two conferences on youtube the last few months.

**Reply by alice-i-cecile** [reply]
Like always, if folks are wondering if Bevy is a good fit for their use case, please reach out to me and I'll give you half an hour of my honest opinion whenever you'd like :)

It's about 50-50 right now!

**Reply by stefanpie** [reply]
Hello, I am also interested in learning more about the strengths of Bevy for visualization for science and engineering-related projects. I am also giving a talk at the workshop this year about electronic design automation/chip design and was curious about the strengths and weaknesses of using Bevy for visualization of the data and algorithms used in this domain. Most of my visualizations would be 2D-based.

I think my major concern is that I want to be able to quickly prototype something in Bevy I can plug into my existing code/prototypes without having to make invasive modifications to my code, as well as a quick way to make a GUI that goes along with the main rendering.

I have yet to look into the entire Bevy ecosystem, but if you have suggestions or pointers, I would love to hear.

**Reply by alice-i-cecile** [reply]
@colepoirier has been looking at chip design in Bevy for a number of years, and may have some more specific advice :)

The main strategy I would use for integration into existing code for visualization is to extract it into relatively monolithic objects, then transform those into ECS state. From there, start with spawning e.g. one sprite per thing you want to visualize, and only worry about optimizing your design further if the performance isn't satisfactory.

For quick GUIs, I would 100% use egui: it's perfect for this sort of knobs-and-buttons tweaking. Take the output from your GUI state, diff it against the last one, and then run the external simulation as needed with new parameters. CLI input for params + file writing results is a really flexible and simple way to get this workflow off the ground.

**Reply by ScottBoyce** [reply]
@alice-i-cecile Not sure how to send direct messages on here so I emailed you at your gmail account. Hopefully that is ok. Take your time with responding, just wanted to double check if you got my email.
